crates/bevy_entropy/src/component.rs

```diff
@@ -0,0 +1,269 @@
+
+use crate::{
+    resource::GlobalEntropy,
+    thread_local_entropy::ThreadLocalEntropy,
+    traits::{EntropySource, SeedableEntropySource},
+};
+use bevy_ecs::{prelude::Component, system::ResMut, world::Mut};
+use rand_core::{RngCore, SeedableRng};
+
+#[cfg(feature = "serialize")]
+use serde::{Deserialize, Serialize};
+
+#[cfg(feature = "bevy_reflect")]
+use bevy_reflect::{FromReflect, Reflect, ReflectDeserialize, ReflectSerialize};
+
+/// An [`EntropyComponent`] that wraps a random number generator that implements
+/// [`RngCore`] & [`SeedableRng`].
+///
+/// ## Creating new [`EntropyComponent`]s.
+///
+/// You can creates a new [`EntropyComponent`] directly from anything that implements
+/// [`RngCore`] or provides a mut reference to [`RngCore`], such as [`ResMut`] or a
+/// [`Component`], or from a [`RngCore`] source directly.
+///
+/// ## Examples
+///
+/// Randomised Component:
+/// ```
+/// use bevy_ecs::{
+///     prelude::Component,
+///     system::Commands,
+/// };
+/// use bevy_entropy::prelude::*;
+/// use rand_chacha::ChaCha8Rng;
+///
+/// #[derive(Component)]
+/// struct Source;
+///
+/// fn setup_source(mut commands: Commands) {
+///     commands
+///         .spawn((
+///             Source,
+///             EntropyComponent::<ChaCha8Rng>::default(),
+///         ));
+/// }
+/// ```
+///
+/// Seeded from a resource:
+/// ```
+/// use bevy_ecs::{
+///     prelude::{Component, ResMut},
+///     system::Commands,
+/// };
+/// use bevy_entropy::prelude::*;
+/// use rand_chacha::ChaCha8Rng;
+///
+/// #[derive(Component)]
+/// struct Source;
+///
+/// fn setup_source(mut commands: Commands, mut global: ResMut<GlobalEntropy<ChaCha8Rng>>) {
+///     commands
+///         .spawn((
+///             Source,
+///             EntropyComponent::from(&mut global),
+///         ));
+/// }
+/// ```
+///
+/// Seeded from a component:
+/// ```
+/// use bevy_ecs::{
+///     prelude::{Component, Query, With, Without},
+///     system::Commands,
+/// };
+/// use bevy_entropy::prelude::*;
+/// use rand_chacha::ChaCha8Rng;
+///
+/// #[derive(Component)]
+/// struct Npc;
+/// #[derive(Component)]
+/// struct Source;
+///
+/// fn setup_npc_from_source(
+///    mut commands: Commands,
+///    mut q_source: Query<&mut EntropyComponent<ChaCha8Rng>, (With<Source>, Without<Npc>)>,
+/// ) {
+///    let mut source = q_source.single_mut();
+///
+///    for _ in 0..2 {
+///        commands
+///            .spawn((
+///                Npc,
+///                EntropyComponent::from(&mut source)
+///            ));
+///    }
+/// }
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq, Component)]
+#[cfg_attr(feature = "bevy_reflect", derive(Reflect, FromReflect))]
+#[cfg_attr(feature = "serialize", derive(Serialize, Deserialize))]
+#[cfg_attr(
+    feature = "serialize",
+    serde(bound(deserialize = "R: for<'a> Deserialize<'a>"))
+)]
+#[cfg_attr(
+    all(feature = "serialize", feature = "bevy_reflect"),
+    reflect_value(Debug, PartialEq, Serialize, Deserialize)
```

Comment:
This should likely also register the following type data: `ReflectFromReflect` and (if we want this to be used in scenes) `ReflectComponent`.